File: src/domain/entities/item-pedido.ts
import { ItemCarrinho } from "./item-carrinho.js";
import { Produto } from "./produto.js";

export type ItemPedidoProps = {
  produtoId: number;
  nomeProduto: string;
  quantidade: number;
  precoUnitario: number;
  subtotal: number;
};

export class ItemPedido {
  readonly produtoId: number;
  readonly nomeProduto: string;
  readonly quantidade: number;
  readonly precoUnitario: number;
  readonly subtotal: number;

  private constructor(props: ItemPedidoProps) {
    this.produtoId = props.produtoId;
    this.nomeProduto = props.nomeProduto;
    this.quantidade = props.quantidade;
    this.precoUnitario = props.precoUnitario;
    this.subtotal = props.subtotal;
  }

  static criarSnapshot(item: ItemCarrinho, produto: Produto): ItemPedido {
    return new ItemPedido({
      produtoId: produto.id,
      nomeProduto: produto.nome,
      quantidade: item.quantidade,
      precoUnitario: produto.preco,
      subtotal: item.quantidade * produto.preco
    });
  }
}

File: claude/backend/src/main/java/com/ecommerce/domain/ItemPedido.java
package com.ecommerce.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;

/**
 * Snapshot historico e imutavel de um produto no momento da compra (RN01).
 *
 * <p>Congela {@code produtoId}, {@code nomeProduto}, {@code quantidade},
 * {@code precoUnitario} e {@code subtotal}. Alteracoes futuras no produto nao
 * afetam pedidos antigos (RN04).</p>
 */
@Entity
@Table(name = "item_pedido")
public class ItemPedido {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(optional = false)
    @JoinColumn(name = "pedido_id", nullable = false)
    private Pedido pedido;

    @Column(name = "produto_id", nullable = false)
    private Long produtoId;

    @Column(name = "nome_produto", nullable = false)
    private String nomeProduto;

    @Column(nullable = false)
    private Integer quantidade;

    @Column(name = "preco_unitario", nullable = false, precision = 12, scale = 2)
    private BigDecimal precoUnitario;

    @Column(nullable = false, precision = 12, scale = 2)
    private BigDecimal subtotal;

    protected ItemPedido() {
        // exigido pelo JPA
    }

    private ItemPedido(Long produtoId, String nomeProduto, Integer quantidade, BigDecimal precoUnitario) {
        this.produtoId = produtoId;
        this.nomeProduto = nomeProduto;
        this.quantidade = quantidade;
        this.precoUnitario = precoUnitario;
        this.subtotal = precoUnitario.multiply(BigDecimal.valueOf(quantidade));
    }

    /**
     * Cria o snapshot imutavel a partir de um item do carrinho e do produto
     * correspondente, congelando nome e preco no momento da compra (RN01).
     */
    public static ItemPedido criarSnapshot(ItemCarrinho item, Produto produto) {
        return new ItemPedido(
                produto.getId(),
                produto.getNome(),
                item.getQuantidade(),
                item.getPrecoUnitario());
    }

    void associarPedido(Pedido pedido) {
        this.pedido = pedido;
    }

    public Long getId() {
        return id;
    }

    public Long getProdutoId() {
        return produtoId;
    }

    public String getNomeProduto() {
        return nomeProduto;
    }

    public Integer getQuantidade() {
        return quantidade;
    }

    public BigDecimal getPrecoUnitario() {
        return precoUnitario;
    }

    public BigDecimal getSubtotal() {
        return subtotal;
    }
}
